### PR TITLE
Add all college support and autocomplete dropdown

### LINKED ISSUE #24: Add autocomplete dropdown to main page
A dropdown like [here](https://www.w3schools.com/howto/howto_js_autocomplete.asp) would allow users to pick from many different universities without scrolling a bunch.

### LINKED ISSUE #55: Add Multiple College Location Functionality
We currently only support our own colleges but we would like to have all US colleges included sorted alphabetically! 

image

### PR DESCRIPTION
**Pull request by mirrorkeydev**
#### Add all-college support (@mirrorkeydev, @areetaw)
By replacing the JSON file containing three colleges to one containing every single US college (~7000), we can add support for pretty much all US college students. The JSON file has had unneeded fields removed, been alphabetized by college name, and been flattened in order to optimize performance. Resolves #55.

Regarding impact on page load, the overall impact of changing from a 2kB JSON file to a 700kB JSON file is fairly negligible in this case. This is mainly due to the way that the file is being loaded, which is completely asynchronous and independent from the rest of the page's elements. I did some quick tests just to confirm this is the case: though the file is 35000% larger than before, the average DOMContentLoaded and Load times (n = 10 for each JSON) hardly changed.

#### Add autocomplete dropdown (@mirrorkeydev)
Adds an autocomplete dropdown to the home page and the sign up page using the `datalist` element. Resolves #24. Options must be added in a very particular way (using document fragments and temporarily removing the ID) to resolve performance issues, see below:
https://github.com/googleinterns/noms/blob/49cb8de302d0a92ae04670f8e41ade1e280f70d1/src/main/webapp/assets/js/indexScript.js#L34-L52
([src1](https://stackoverflow.com/questions/45083176/is-this-a-chrome-ui-performance-bug-related-to-input-datalist), [src2](https://stackoverflow.com/questions/30336489/how-to-populate-a-large-datalist-2000-items-from-a-dictionary), [src3](https://stackoverflow.com/questions/29882361/show-datalist-labels-but-submit-the-actual-value/29882539))
If the user enters a college not on the list (as it is a textbox, and you can technically type anything), and hits `enter`, no navigation occurs. Users are thus limited to the colleges in the JSON.
image
image
image

#### Add all college support to sign up page (@areetaw)
Adds a JS file for sign up page in order to add multiple colleges autocomplete drop-down. When users subscribe/unsubscribe, the college that they pick is now set to be a unique ID in the backend rather than a string of the university name.

image

### CODE REVIEW

File: src/main/webapp/assets/css/indexStyle.css [hunk 1]
```diff
@@ -39,33 +39,37 @@
...
   margin: 0;
 }
 
-select:invalid {
-  color: gray;
+#colleges-input {
+  background: #dddbdb;
+  border: 2px solid #dddbdb;
+  box-sizing: border-box;
+  border-radius: 5px;
```
Comment (mrboxtobox) [inline src/main/webapp/assets/css/indexStyle.css:50]: Fix ordering

File: src/main/webapp/assets/js/signUpScript.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+async function onLoad() {
+  const collegeLocations = await (await fetch('./assets/college-locations.json')).json();
+
+  // Grab the datalist and remove its ID (destroying the select-datalist relationship),
+  // to improve performance while adding the options to the datalist.
+  const collegeDataList = document.getElementById('colleges');
+  collegeDataList.removeAttribute('id');
+
+  // Add all colleges as datalist options. We use document fragments because the
+  // DOM is slow if we add each option individually and let the DOM update in between.
```
Comment (mrboxtobox) [inline src/main/webapp/assets/js/signUpScript.js:41]: These are very good comments, explaining the `why`.

File: src/main/webapp/assets/js/signUpScript.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  const frag = document.createDocumentFragment();
```
Comment (mrboxtobox) [inline src/main/webapp/assets/js/signUpScript.js:42]: fragment

File: src/main/webapp/assets/js/signUpScript.js (hunk 2 above)
Comment (mirrorkeydev) [reply src/main/webapp/assets/js/signUpScript.js:41]: Thank you! 😁

File: src/main/webapp/assets/college-locations.json [hunk 4]
```diff
@@ -1,77 +1 @@
...
...,"LAT":"46.3748","LON":"-120.394972"},{"UNITID":"431275","NAME":"Herkimer County BOCES-Practical Nursing Program","LAT":"43.015797","LON":"-75.029952"},{"UNITID":"191612","NAME":"Herkimer County Community College","LAT":"43.032738","LON":"-75.007206"},{"UNITID":"200785","NAME":"Herzing University-Akron","LAT":"41.02817","LON":"-81.493109"},{"UNITID":"140340","NAME":"Herzing University-Atlanta","LAT":"33.754218","LON":"-84.387251"},{"UNITID":"101365","NAME":"Herzing University-Birmingham","LAT":"33.468466","LON":"-86.832465"},{"UNITID":"459851","NAME":"Herzing University-Brookfield","LAT":"43.035738","LON":"-88.110994"},{"UNITID":"433536","NAME":"Herzing University-Kenner","LAT":"30.001605","LON":"-90.239525"},{"UNITID":"459842","NAME":"Herzing University-Kenosha","LAT":"42.603889","LON":"-87.857452"},{"UNITID":"240392","NAME":"Herzing University-Madison","LAT":"43.14415","LON":"-89.291561"},{"UNITID":"174154","NAME":"Herzing University-Minneapolis","LAT":"44.975225","LON":"-93.404863"},{"UNITID":"457022","NAME":"Herzing University-Toledo","LAT":"41.639266","LON":"-83.665453"},{"UNITID":"386472","NAME":"Herzing University-Winter Park","LAT":"28.615784","LON":"-81.312669"},{"UNITID":
```
Comment (mrboxtobox) [inline src/main/webapp/assets/college-locations.json:1]: Add newline.

File: src/main/webapp/assets/css/indexStyle.css (hunk 1 above)
Comment (mirrorkeydev) [reply src/main/webapp/assets/css/indexStyle.css:50]: Gotcha!

File: src/main/webapp/assets/js/signUpScript.js (hunk 3 above)
Comment (mirrorkeydev) [reply src/main/webapp/assets/js/signUpScript.js:42]: 👍

File: src/main/webapp/assets/college-locations.json (hunk 4 above)
Comment (mirrorkeydev) [reply src/main/webapp/assets/college-locations.json:1]: Will do!

**Comment by mirrorkeydev** [comment]
Thank you Sanya! I will get on adding these things to the doc, and will conduct some page load analysis :)

File: src/main/webapp/assets/css/mainStyle.css [hunk 5]
```diff
@@ -199,10 +199,11 @@ header {
   padding: 1rem;
 }
 
+#colleges-input-form,
```
Comment (ubahl) [inline src/main/webapp/assets/css/mainStyle.css:202]: oh cool I didn't know you could do this!

File: src/main/webapp/assets/js/indexScript.js [hunk 6]
```diff
@@ -31,25 +31,42 @@ document.addEventListener('DOMContentLoaded', onLoad);
...
 
-  // Add all colleges as dropdown options
-  const collegeDropdown = document.getElementById('colleges-select');
+  // Grab the datalist and remove its ID (destroying the select-datalist relationship),
+  // to improve performance while adding the options to the datalist.
+  const collegeDataList = document.getElementById('colleges');
+  collegeDataList.removeAttribute('id');
+
+  // Add all colleges as datalist options. We use a document fragment because the
+  // DOM is slow if we add each option individually and let the DOM update in between.
```
Comment (ubahl) [inline src/main/webapp/assets/js/indexScript.js:40]: Cool!

**Comment by areetaw** [comment]
@ubahl wanted to resolve @mrboxtobox comments so she'll be on implementing changes on `PostDataServlet.java` and `Post.java`!

**Comment by mirrorkeydev** [comment]
@mrboxtobox Regarding impact on page load, the overall impact of changing from a 2kB JSON file to a 700kB JSON file is fairly negligible in this case. This is mainly due to the way that the file is being loaded, which is completely asynchronously and independently from the rest of the page's elements. I did some quick tests just to confirm this is the case: though the file is 35000% larger than before, the average DOMContentLoaded and Load times (n = 10 for each JSON) hardly changed. I will add this info to the body of the PR as well.

File: src/main/webapp/assets/js/indexScript.js [hunk 7]
```diff
@@ -31,25 +31,42 @@ document.addEventListener('DOMContentLoaded', onLoad);
...
 }
 
 /**
  * When the user selects a college from the dropdown, we immediately
  * navigate them to the appropriate college's page.
  */
 function navigateUserToCollegePage() {
-  const collegeid = document.getElementById('colleges-select').value;
-  window.location.href = `/find-events.html?collegeid=${collegeid}`;
+  const collegename = document.getElementById('colleges-input').value;
```
Comment (areetaw) [inline src/main/webapp/assets/js/indexScript.js:63]: camelCase!

File: src/main/webapp/assets/js/indexScript.js [hunk 8]
```diff
@@ -31,25 +31,42 @@ document.addEventListener('DOMContentLoaded', onLoad);
...
+  const option = document.querySelector(`#colleges option[value='${collegename}']`);
+
+  // Only navigate to the page if that college exists in our list of colleges.
+  // TODO: display 'We haven't heard of that college!"/similar to the user if not recognized.
+  if (option) {
+    const collegeid = option.dataset.value;
```
Comment (areetaw) [inline src/main/webapp/assets/js/indexScript.js:69]: here as well!

File: src/main/webapp/assets/js/indexScript.js [hunk 9]
```diff
@@ -31,25 +31,42 @@ document.addEventListener('DOMContentLoaded', onLoad);
...
 
 /**
  * When the user selects a college from the dropdown, we immediately
  * navigate them to the appropriate college's page.
  */
 function navigateUserToCollegePage() {
-  const collegeid = document.getElementById('colleges-select').value;
-  window.location.href = `/find-events.html?collegeid=${collegeid}`;
+  const collegename = document.getElementById('colleges-input').value;
+  const option = document.querySelector(`#colleges option[value='${collegename}']`);
```
Comment (areetaw) [inline src/main/webapp/assets/js/indexScript.js:64]: here too!

**Pull request merged**